daemon/src/oracle.rs

```diff
@@ -0,0 +1,332 @@
+use anyhow::Result;
+use cfd_protocol::secp256k1_zkp::{schnorrsig, SecretKey};
+use rocket::time::{format_description, Duration, OffsetDateTime, Time};
+use std::convert::{Infallible, TryInto};
+
+/// Where `olivia` is located.
+const OLIVIA_URL: &str = "https://h00.ooo/";
+
+const OLIVIA_EVENT_TIME_FORMAT: &str = "[year]-[month]-[day]T[hour]:[minute]:[second]";
+
+pub async fn watch_next_announcement<A>(actor_address: xtra::Address<A>) -> Result<()>
+where
+    A: xtra::Handler<Announcements>,
+{
+    tokio::spawn(async move {
+        let backoff = backoff::ExponentialBackoff {
+            max_elapsed_time: None,
+            ..backoff::ExponentialBackoff::default()
+        };
+
+        let res = backoff::future::retry::<Infallible, anyhow::Error, _, _, _>(backoff, || {
+            let actor_address = actor_address.clone();
+            async move {
+                let mut announcements = Vec::new();
+                loop {
+                    let mut new_announcements = Vec::new();
+                    for event_url in next_urls().iter() {
+                        let res = reqwest::get(event_url).await.map_err(|_| {
+                            backoff::Error::Transient(anyhow::anyhow!("request failed"))
+                        })?;
+
+                        let announcement = res.json::<Announcement>().await.map_err(|_| {
+                            backoff::Error::Permanent(anyhow::anyhow!(
+                                "failed to deserialize announcement"
+                            ))
+                        })?;
+
+                        new_announcements.push(announcement)
+                    }
+
+                    if new_announcements != announcements {
+                        announcements = new_announcements.clone();
+                        actor_address
+                            .do_send_async(Announcements(new_announcements))
+                            .await
+                            .expect("receiving actor to be up");
+                    }
+
+                    tokio::time::sleep(std::time::Duration::from_secs(3600)).await;
+                }
+            }
+        })
+        .await;
+
+        if let Err(e) = res {
+            tracing::error!(
+                "Announcement watcher encountered an unrecoverable error: {:#}",
+                e
+            );
+        }
+    });
+
+    Ok(())
+}
+
+pub async fn watch_attestation<CFD, M>(
+    announcement: Announcement,
+    cfd_actor_address: xtra::Address<CFD>,
+    monitor_actor_address: xtra::Address<M>,
+) -> Result<()>
+where
+    CFD: xtra::Handler<Attestation>,
+    M: xtra::Handler<Attestation>,
+{
+    let wait_time =
```

Comment:
:mega: Okay, this looks very wrong. 

I would welcome suggestions on how to handle this. Maybe we shouldn't even wait until the `expected_outcome_time` and just try regularly?